### ISSUE TITLE
Munged SciDB queries

### ISSUE THREAD

**Issue by mcolosimo-p4** [Paradigm4/SciDBR#234]
For sometime, we've been seeing odd failing queries making it to SciDB from long running uploads consisting of multi-part uploads. These are very rare and have been hard to reproduce.

For example:

> store(input(remove(R_array8f1b4bc89bf21631571798133224844)0:17144399,1000,0],'/dev/shm/shim_input_buf_FaFEO9', -2, '(string null)'),R_array8f1b64c9ca111631571798133224844)

The query should be:

> store(input(< chrom : int64 null>  [i=0:17144399,1000,0],'/dev/shm/shim_input_buf_FaFEO9', -2, '(string null)'),R_array8f1b64c9ca111631571798133224844)

The above query is built like:

```R
     query = sprintf("store(input(%s,'%s', -2, '(%s null)'), %s)", load_schema, ans, type, name)
```

So `load_schema` is munged, which is built like:

```R
    schema = sprintf(
        "< %s : %s null>  [%s=%.0f:%.0f,%.0f,%.0f]", attr_name, force_type, dimname, start[[1]],
        nrow(X) - 1 + start[[1]], min(nrow(X), chunkSize), overlap[[1]])
    load_schema = schema
```

From examining previously seen munged queries, it appears that the `sprintf` during formatting the `%0.f` statements is when the *remove* query is placed into the result.

Examining the SciDB logs, we found that the *remove* queries were the previously executed query, one generated when the backing R object holding a handle to the temporary array is GC'ed.

This also is created using `sprintf` in the finalizer, see [utility.R#L61](https://github.com/Paradigm4/SciDBR/blob/e4230c46e5a24f55d43a4462fb6a82bf7ae07240/R/utility.R#L61):

```R
     scidbquery(db, sprintf("remove(%s)", e$name))
```

Reviewing the R code for `sprintf` for R3.6.0, it uses a static variable to hold the output string, see [sprintf.c#L90](https://github.com/wch/r-source/blob/tags/R-3-6-0/src/main/sprintf.c#L90). This makes `sprintf` non-reentrant (even from the same thread). Note, that this seems to be still present in R4, thus we expect the same issue.

It seems that GC it triggered while `sprintf` is building up a string, specifically for `%0.f` formatting.

We were able to create a small test script that exposes this issue. `paste` has the same potential but we have been unable to trigger a similar behavior using it.  However, we have decided to just move the `sprintf`s outside of the finalizer.

**Comment by ksens** [comment]
Awesome stuff! For future reference, can you paste your repro script here?

**Comment by mcolosimo-p4** [comment]
@ksens here is the repro script. I think the key part is the format conversion of an integer to double; e.g., `sprintf("%.0f" , start[[1]])` which can trigger the GC while in the sprintf C function.

```R
# Example code that shows that sprintf is not reentrant 
# Usually fails rather fast (< 100,000 objects created)
library(uuid)

.GlobalEnv$in_other_sprintf <- F
.GlobalEnv$borked = F

setClass("my_class",
         representation(name="character",
                        data="character",
                        meta="environment"),
         S3methods=TRUE)

get_my_class = function(name){
  string_size = floor(runif(1, min = 5, max = 20))
  num_strings = floor(runif(1, min = 10, max = 50))
  obj = new("my_class", name=name)
  obj@data = stringi::stri_rand_strings(string_size, num_strings)
  obj@meta = new.env()
  obj@meta$name = name
  reg.finalizer(obj@meta, function(e){
    query = sprintf("FINALIZED(%s)", e$name)
    if (.GlobalEnv$in_other_sprintf) {
      .GlobalEnv$borked = T
    }
  })
  obj
}

make.names_ = function(x)
{
  gsub("\\.", "_", make.names(x, unique=TRUE), perl=TRUE)
}

# x is vector of existing values
# y is vector of new values
# returns a set the same size as y with non-conflicting value names
make.unique_ = function(x, y)
{
  z = make.names(gsub("_", ".", c(x, y)), unique=TRUE)
  gsub("\\.", "_", utils::tail(z, length(y)))
}

my_class_count = 1
times_in_sprintf = 0
my_strs <- c()
count <- 0

# try to reproduce SciDBR as close as possible
force_type = "string"
chunkSize = 1000
start = as.integer(c(0, 0))
nrowX = as.integer(17144399)
overlap = c(0, 0)
overlap = as.integer(overlap)

x <- c()
while(T){
  attr_name = UUIDgenerate()
  dimname = make.unique_(attr_name, "i") 
  
  classes_to_make <- floor(runif(1, min = 1, max = 10))
  for (i in my_class_count:my_class_count+classes_to_make) {
    x <- c( x, get_my_class(paste0("object_",i)))
  }
  my_class_count <- my_class_count + classes_to_make
  
  .GlobalEnv$in_other_sprintf <- T
  my_str <- sprintf("< %s : %s null>  [%s=%.0f:%.0f,%.0f,%.0f]", attr_name, force_type, dimname, 
                    start[[1]], nrowX - 1 + start[[1]], min(nrowX, chunkSize), overlap[[1]])
  .GlobalEnv$in_other_sprintf <- F
  
  count <- count + 1
  my_strs <- c(my_strs, my_str)
  if( .GlobalEnv$borked) {
    times_in_sprintf = times_in_sprintf + 1
    if (grepl("FINALIZED", my_str)) {
      print(my_str)
      stop(paste0("FINALIZED in my_str, number of times in sprintf ", times_in_sprintf))
    } 
    print(paste0("GC'ed while in sprintf, but my_str is fine: ", times_in_sprintf, " my_class_count: ", my_class_count, " count: ",count))
    .GlobalEnv$borked = F
  }
  if (count %% 10 == 0) {
    x <- sample(x, floor(runif(1, min=0, max=length(x))), replace=FALSE)
  }
  if (my_class_count %% 10000 == 0 ) {
    if(any(grepl("FINALIZE", my_strs))) {
      stop("'FINALIZE' in my_strs")
    }
    print(paste("my class count: ", my_class_count, "    count: ", count))
  }
}
```

**Comment by ksens** [comment]
Closed via #235 . 

For the benefit of web-crawlers, adding the following issue gist:

- Having `sprintf` in a finalizer code-block can trigger non-reentrant nature of `sprintf` leading to non-deterministic behaviour in application code
- Solution: Do not formulate strings using `sprintf` within finalizers. Formulate them outside finalizer, and then use those strings as needed
